core/src/main/java/org/axonframework/eventsourcing/eventstore/AbstractEventStorageEngine.java

```diff
@@ -108,12 +107,14 @@ public DomainEventStream readEvents(String aggregateIdentifier, long firstSequen
     }
 
     @Override
-    public Optional<DomainEventMessage<?>> readSnapshot(String aggregateIdentifier) {
-        return readSnapshotData(aggregateIdentifier).filter(snapshotFilter).map(entry -> {
-            DomainEventStream stream =
-                    EventUtils.upcastAndDeserializeDomainEvents(Stream.of(entry), serializer, upcasterChain, false);
-            return stream.hasNext() ? stream.next() : null;
-        });
+    public Stream<DomainEventMessage<?>> readSnapshots(String aggregateIdentifier) {
+        return readSnapshotData(aggregateIdentifier).filter(snapshotFilter)
```

Comment:
Given that there is a filter here that should filter out incompatible snapshots, we don't need a SnapshotResolver anymore.

Reply:
We can remove the `SnapshotResolver`, but `snapshotFilter` can return multiple snapshots, `SnapshotResolver` ensures that there will be at most one. 

Is it OK to _filter_ the snapshots and take the first one (even if the filter returns several of them)? 

Anyway, I'll implement the required change, just want to raise a concern here :)

Reply:
Your assumption is correct. The filter defines what a usable snapshot is for the current application, and we then take the first snapshot from the stream

Reply:
Awesome! Already implemented like that ;)